### PR TITLE
HPCC-18041 Unicode Implementation for ExcludeLastWord

### PR DESCRIPTION
**Pull request by dskaff**
- Adds testcases to ecllibarry/teststd/uni/TestExcludeLastWord.ecl
- Adds documentation to ecllibrary/std/Uni.ecl
- Adds code to plugins/unicodelib.cpp

Signed-off-by: David Skaff <david.skaff@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by dskaff** [comment]
@johnholt please code review.

File: plugins/unicodelib/unicodelib.cpp [hunk 1]
```diff
@@ -719,6 +720,42 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+            // Exclude spaces, punctuation, and the like.
+            //   A status value UBRK_WORD_NONE indicates that the boundary does
+            //   not start a word or number.
+            ++wordidx;
+            wordBeginning = bi.previous();
+            if (bi.getRuleStatus() != UBRK_WORD_NONE)
+            {
+                ++wordidx;
+            }
+            if (bi.previous() == bi.first() && wordidx == 1)
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: This isn't safe.  Both bi.first() and bi.previous() modify the current iterator position.  It is undefined which is called first.

I think you should be comparing against 0.

Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: It isn't clear what this code is here for - there should be a comment to describe why

File: plugins/unicodelib/unicodelib.cpp [hunk 2]
```diff
@@ -719,6 +720,42 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+void excludeLastWord(RuleBasedBreakIterator& bi, UnicodeString & source)
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: style: Not keen on a parameter that is modified being called source (see other PRs for comments)

File: plugins/unicodelib/unicodelib.cpp (hunk 1 above)
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: Actually documentation says:  
The character index of the preceding text boundary, or UBRK_DONE if all text boundaries have been returned.

I'm not sure what you are testing here.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@dskaff a good first stab, but I have some concerns about some of the code.

**Comment by ghalliday** [comment]
Also checkboxes are not completed on the PR.

File: plugins/unicodelib/unicodelib.cpp (hunk 1 above)
Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: Normally, the function only removes the trailing white spaces of the last word, but when the source string only contains one word, the leading and trailing white spaces are removed (according to the string test cases).
The condition sets the iterator at the position previous to the beginning of the last word. If the string is only one word, then that position should be equal to bi.first().

Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: I'll change bi.first() to 0

Comment (dskaff) [inline plugins/unicodelib/unicodelib.cpp]: I don't believe that it will return UBRK_DONE unless you are using bi.preceding() while on the first boundary.

**Comment by dskaff** [comment]
@ghalliday please code review

File: plugins/unicodelib/unicodelib.cpp [hunk 3]
```diff
@@ -719,6 +720,39 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+    {
+        int breakType = bi.getRuleStatus();
+        if (breakType != UBRK_WORD_NONE)
+        {
+            // Exclude spaces, punctuation, and the like.
+            //   A status value UBRK_WORD_NONE indicates that the boundary does
+            //   not start a word or number.
+            ++wordidx;
+            wordBeginning = bi.previous();
+            if (bi.getRuleStatus() != UBRK_WORD_NONE)
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: This needs a comment explaining the logic - I can't follow it - or why bi.previous() would be called twice before checking it is 0.

**Comment by dskaff** [comment]
@ghalliday I've commented the code

**Review by ghalliday — APPROVED** [review]

**Pull request merged**